### ISSUE TITLE
distsql: update HLC when distsql messages are exchanged

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#7526]
With distsql, hosts might not talk to each other through kv/dist_sender, but only through distsql. So we need the distsql layer to update the logical clocks.

For streams, it's not clear if we need to send and update on every message, or it would be sufficient to do it only once per stream (in the header).

CC @andreimatei 

Jira issue: CRDB-6177

**Comment by RaduBerinde** [comment]
CC @tschottdorf - any words of wisdom?

**Comment by andreimatei** [comment]
One important thing is for a node `n` to not do a read `r` with a time higher than the node's clock. Because, if the node transfers its lease away, the next lease starts at `n`'s clock. And now `r` is not part of the new leader's read cache.

**Comment by tschottdorf** [comment]
It depends. Updating for every message in the stream is definitely safer, but also not necessary. For example, when streaming a result set, it's not necessary to exchange HLC timestamps all the time.

What's a concrete example where DistSender would be bypassed?

**Comment by RaduBerinde** [comment]
So say a node A receives a query and it knows the relevant ranges are lead by B and C. It will set up a distsql flow between B and C.  B and C will do KV operations to their respective ranges. Assuming replicas are on other nodes, all communication between A, B, C happens through distsql.

**Comment by tschottdorf** [comment]
But while you set up that flow, the ranges may move, so it seems that that node which is assumed to have the range locally would still have to use `DistSender` (or something like it) to potentially fetch the data (which would almost always be a simple pass-through, but could require the full power of DistSender as it exists today).

**Comment by tschottdorf** [comment]
oh - sorry, I misunderstood what you are saying. I have to think about it a little more, but I think the biggest concern here is read-what-you-wrote, which means that what matters is that all nodes have an update from each touched node's HLC by the time the operation is finished.

**Comment by RaduBerinde** [comment]
#9619 is somewhat related. We may be able to implement a general mechanism that exchanges and updates the logical clocks on all RPCs, using gRPC interceptors.

**Comment by jordanlewis** [comment]
Is this still a problem? Does it have a user-visible consequence?

**Comment by andreimatei** [comment]
We've started marshaling observed timestamps from TableReaders to the gateway node through the `ProducerMetadata.TxnCoordMeta.Transaction.ObservedTimestamps`. But that's a subcase for what this issues is saying. We should opportunistically propagate and record observed timestamps between all nodes that communicate.
Consequences - smaller uncertainty windows.

**Comment by andreimatei** [comment]
I've started working on propagating clock signals on all RPCs here:
https://github.com/andreimatei/cockroach/tree/grpc.clock-interceptors

**Comment by jordanlewis** [comment]
Move this to Core since @andreimatei is working on this - feel free to close if there's a better issue to capture the work.
